tpls/desul/include/desul/atomics/Compare_Exchange_OpenACC.hpp

```diff
@@ -0,0 +1,266 @@
+/*
+Copyright (c) 2019, Lawrence Livermore National Security, LLC
+and DESUL project contributors. See the COPYRIGHT file for details.
+Source: https://github.com/desul/desul
+
+SPDX-License-Identifier: (BSD-3-Clause)
+*/
+
+#ifndef DESUL_ATOMICS_COMPARE_EXCHANGE_OPENACC_HPP_
+#define DESUL_ATOMICS_COMPARE_EXCHANGE_OPENACC_HPP_
+
+#include <openacc.h>
+
+#include <desul/atomics/Common.hpp>
+#include <desul/atomics/Thread_Fence_OpenACC.hpp>
+
+namespace desul {
+namespace Impl {
+
+#pragma acc routine seq
+template <class T, class MemoryOrder, class MemoryScope>
+std::enable_if_t<!std::is_arithmetic<T>::value, T> device_atomic_exchange(T* dest, T value, MemoryOrder, MemoryScope scope) {
+  if (acc_on_device(acc_device_not_host)) {
+    printf("Kokkos Error in device_atomic_exchange(): Not supported atomic "
+                  "operation in the OpenACC backend\n");
+  }
+  //FIXME_OPENACC OpenACC lock APIs are not implemented.
+  // Acquire a lock for the address
+  //while (!lock_address_openacc((void*)dest, scope)) {
+  //}
+  //device_atomic_thread_fence(MemoryOrderAcquire(), scope);
+  T return_val = *dest;
+  *dest = value;
+  //device_atomic_thread_fence(MemoryOrderRelease(), scope);
+  //unlock_address_openacc((void*)dest, scope);
+  return return_val;
+}
+
+#ifdef KOKKOS_COMPILER_NVHPC
```

Comment:
You might want to add a comment or define a macro that express better that we will be using CUDA atomic functions in that case.

Reply:
Added a comment.

Reply:
Changed `KOKKOS_COMPILER_NVHPC` to `__NVCOMPILER`.